[Iceberg] `SELECT` fails after nested field schema update with Parquet format

Hi, 

We have a table with field `top_level array(ROW(a varchar, b varchar))`
When change type of nested ROW to `top_level array(ROW(c varchar))` we can't query this field

Tested on Trino 453 and latest master

Query fails with:
```java
Caused by: io.trino.spi.TrinoException: Failed to read Parquet file: local:///test_schema/test_table_schema-b9cc55871a7f425fb7388f5f59a38050/data/20240802_143046_00008_ayge4-3f8d3cd1-fb4a-4b17-a647-1f49a3b0ecc9.parquet
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.handleException(IcebergPageSourceProvider.java:1459)
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.lambda$createParquetPageSource$26(IcebergPageSourceProvider.java:1024)
	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:75)
	at io.trino.spi.block.LazyBlock$LazyData.load(LazyBlock.java:312)
	at io.trino.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:291)
	at io.trino.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:186)
	at io.trino.spi.Page.getLoadedPage(Page.java:238)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:271)
	at io.trino.operator.Driver.processInternal(Driver.java:403)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:306)
	at io.trino.operator.Driver.tryWithLock(Driver.java:709)
	at io.trino.operator.Driver.process(Driver.java:298)
	at io.trino.operator.Driver.processForDuration(Driver.java:269)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:890)
	at io.trino.execution.executor.dedicated.SplitProcessor.run(SplitProcessor.java:77)
	at io.trino.execution.executor.dedicated.TaskEntry$VersionEmbedderBridge.lambda$run$0(TaskEntry.java:191)
	at io.trino.$gen.Trino_testversion____20240802_143035_71.run(Unknown Source)
	at io.trino.execution.executor.dedicated.TaskEntry$VersionEmbedderBridge.run(TaskEntry.java:192)
	at io.trino.execution.executor.scheduler.FairScheduler.runTask(FairScheduler.java:168)
	at io.trino.execution.executor.scheduler.FairScheduler.lambda$submit$0(FairScheduler.java:155)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:76)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:143)
	at io.trino.parquet.reader.ParquetReader.readArray(ParquetReader.java:340)
	at io.trino.parquet.reader.ParquetReader.readColumnChunk(ParquetReader.java:540)
	at io.trino.parquet.reader.ParquetReader.readBlock(ParquetReader.java:526)
	at io.trino.parquet.reader.ParquetReader.lambda$nextPage$3(ParquetReader.java:252)
	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:72)
	... 24 more
```

### Test to reproduce issue:
```java
    @Test
    public void testReadTableAfterSchemaUpdate()
    {
        assertQuerySucceeds("CREATE TABLE iceberg.test_schema.test_table_schema (top_level array(ROW(a varchar, b varchar)))");

        assertQuerySucceeds("INSERT INTO iceberg.test_schema.test_table_schema select CAST(array[ROW('a', 'b')] AS array(ROW(a varchar, b varchar)))");

        assertQuerySucceeds("alter table iceberg.test_schema.test_table_schema drop column top_level.element.a");
        assertQuerySucceeds("alter table iceberg.test_schema.test_table_schema add column top_level.element.c varchar");
        assertQuerySucceeds("alter table iceberg.test_schema.test_table_schema drop column top_level.element.b");

        assertQuerySucceeds("INSERT INTO iceberg.test_schema.test_table_schema select CAST(array[ROW('c')] AS array(ROW(c varchar)))");

        assertQuerySucceeds("SELECT * from iceberg.test_schema.test_table_schema ");
    }
```

--- a/lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java
+++ b/lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java
@@ -337,7 +337,11 @@ public class ParquetReader
     {
         List<Type> parameters = field.getType().getTypeParameters();
         checkArgument(parameters.size() == 1, "Arrays must have a single type parameter, found %s", parameters.size());
-        Field elementField = field.getChildren().get(0).get();
+        Optional<Field> children = field.getChildren().get(0);
+        if (children.isEmpty()) {
+            return new ColumnChunk(field.getType().createNullBlock(), new int[] {}, new int[] {});
+        }
+        Field elementField = children.get();
         ColumnChunk columnChunk = readColumnChunk(elementField);
 
         ListColumnReader.BlockPositions collectionPositions = calculateCollectionOffsets(field, columnChunk.getDefinitionLevels(), columnChunk.getRepetitionLevels());
@@ -355,7 +359,8 @@ public class ParquetReader
 
         ColumnChunk columnChunk = readColumnChunk(field.getChildren().get(0).get());
         blocks[0] = columnChunk.getBlock();
-        blocks[1] = readColumnChunk(field.getChildren().get(1).get()).getBlock();
+        Optional<Field> valueField = field.getChildren().get(1);
+        blocks[1] = valueField.isPresent() ? readColumnChunk(valueField.get()).getBlock() : parameters.get(1).createNullBlock();
         ListColumnReader.BlockPositions collectionPositions = calculateCollectionOffsets(field, columnChunk.getDefinitionLevels(), columnChunk.getRepetitionLevels());
         Block mapBlock = ((MapType) field.getType()).createBlockFromKeyValue(collectionPositions.isNull(), collectionPositions.offsets(), blocks[0], blocks[1]);
         return new ColumnChunk(mapBlock, columnChunk.getDefinitionLevels(), columnChunk.getRepetitionLevels());

--- a/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java
+++ b/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java
@@ -8539,6 +8539,45 @@ public abstract class BaseIcebergConnectorTest
         }
     }
 
+    @Test // regression test for https://github.com/trinodb/trino/issues/22922
+    void testArrayElementChange()
+    {
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_array_schema_change",
+                "(col array(row(a varchar, b varchar)))",
+                List.of("CAST(array[row('a', 'b')] AS array(row(a varchar, b varchar)))"))) {
+            assertUpdate("ALTER TABLE " + table.getName() + " DROP COLUMN col.element.a");
+            assertUpdate("ALTER TABLE " + table.getName() + " ADD COLUMN col.element.c varchar");
+            assertUpdate("ALTER TABLE " + table.getName() + " DROP COLUMN col.element.b");
+
+            String expected = format == ORC ? "CAST(array[row(NULL)] AS array(row(c varchar)))" : "CAST(NULL AS array(row(c varchar)))";
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES " + expected);
+        }
+    }
+
+    // MAP type is tested in TestIcebergV2.testMapValueSchemaChange
+
+    @Test
+    void testRowFieldChange()
+    {
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_row_schema_change",
+                "(col row(a varchar, b varchar))")) {
+            assertUpdate("INSERT INTO " + table.getName() + " SELECT CAST(row('a', 'b') AS row(a varchar, b varchar))", 1);
+
+            assertUpdate("ALTER TABLE " + table.getName() + " DROP COLUMN col.a");
+            assertUpdate("ALTER TABLE " + table.getName() + " ADD COLUMN col.c varchar");
+            assertUpdate("ALTER TABLE " + table.getName() + " DROP COLUMN col.b");
+
+            String expected = format == ORC || format == AVRO ? "CAST(row(NULL) AS row(c varchar))" : "CAST(NULL AS row(c varchar))";
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("SELECT " + expected);
+        }
+    }
+
     @Test
     public void testObjectStoreLayoutEnabledAndDataLocation()
             throws Exception

--- a/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java
+++ b/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java
@@ -1467,6 +1467,31 @@ public class TestIcebergV2
     }
 
     @Test
+    void testMapValueSchemaChange()
+    {
+        testMapValueSchemaChange("PARQUET", "map(array[1], array[NULL])");
+        testMapValueSchemaChange("ORC", "map(array[1], array[row(NULL)])");
+        testMapValueSchemaChange("AVRO", "NULL");
+    }
+
+    private void testMapValueSchemaChange(String format, String expectedValue)
+    {
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_map_value_schema_change",
+                "WITH (format = '" + format + "') AS SELECT CAST(map(array[1], array[row(2)]) AS map(integer, row(field integer))) col")) {
+            Table icebergTable = loadTable(table.getName());
+            icebergTable.updateSchema()
+                    .addColumn("col.value", "new_field", Types.IntegerType.get())
+                    .deleteColumn("col.value.field")
+                    .commit();
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .as("Format: %s", format)
+                    .matches("SELECT CAST(" + expectedValue + " AS map(integer, row(new_field integer)))");
+        }
+    }
+
+    @Test
     public void testUpdateAfterEqualityDelete()
             throws Exception
     {

